File: examples/t5_generate.cc
/*! @copyright (c) 2022-present Skolkovo Institute of Science and Technology
 *                              (Skoltech), Russia. All rights reserved.
 *                 2023-present Artificial Intelligence Research Institute
 *                              (AIRI), Russia. All rights reserved.
 *
 * NNTile is software framework for fast training of big neural networks on
 * distributed-memory heterogeneous systems based on StarPU runtime system.
 *
 * @file examples/t5_generate.cc
 * Autoregressive text generation with T5 encoder-decoder.
 *
 * Workflow (two-step):
 *
 *   # Step 1 – Convert HF checkpoint and tokenize (Python):
 *   python examples/t5_generate.py \
 *       --model google-t5/t5-small \
 *       --output-dir /tmp/nntile_t5 \
 *       --encoder-prompt "translate English to German: The house is wonderful."
 *
 *   # Step 2 – Run autoregressive generation (this binary):
 *   ./t5_generate \
 *       --config /tmp/nntile_t5/config.json \
 *       --weights /tmp/nntile_t5/weights.safetensors \
 *       --encoder-ids "$(cat /tmp/nntile_t5/encoder_ids.txt)" \
 *       --decoder-ids "$(cat /tmp/nntile_t5/decoder_ids.txt)" \
 *       --max-tokens 32
 *
 * @version 1.1.0
 * */

#include <algorithm>
#include <cstdint>
#include <cstdlib>
#include <cstring>
#include <fstream>
#include <iostream>
#include <map>
#include <sstream>
#include <string>
#include <sys/wait.h>
#include <unistd.h>
#include <vector>

#include <nntile.hh>
#include <nntile/graph/io/safetensors.hh>
#include <nntile/graph/model/t5/t5_for_conditional_generation.hh>
#include <nntile/graph/model/t5/t5_config.hh>
#include <nlohmann/json.hpp>

using namespace nntile;
using namespace nntile::graph;
using namespace nntile::model::t5;
using json = nlohmann::json;

// ── CLI helpers ──────────────────────────────────────────────────────────

struct Args
{
    std::string config_path;
    std::string weights_path;
    std::string encoder_ids_str;
    std::string decoder_ids_str;
    std::string model;
    std::string encoder_prompt;
    std::string output_dir = "/tmp/nntile_t5";
    int max_tokens = 32;
};

static Args parse_args(int argc, char** argv)
{
    Args a;
    for(int i = 1; i < argc; ++i)
    {
        std::string arg = argv[i];
        if(arg == "--config" && i + 1 < argc)
            a.config_path = argv[++i];
        else if(arg == "--weights" && i + 1 < argc)
            a.weights_path = argv[++i];
        else if(arg == "--encoder-ids" && i + 1 < argc)
            a.encoder_ids_str = argv[++i];
        else if(arg == "--decoder-ids" && i + 1 < argc)
            a.decoder_ids_str = argv[++i];
        else if((arg == "--model" || arg == "--model-dir") && i + 1 < argc)
            a.model = argv[++i];
        else if(arg == "--encoder-prompt" && i + 1 < argc)
            a.encoder_prompt = argv[++i];
        else if(arg == "--output-dir" && i + 1 < argc)
            a.output_dir = argv[++i];
        else if(arg == "--max-tokens" && i + 1 < argc)
            a.max_tokens = std::stoi(argv[++i]);
        else if(arg == "--help" || arg == "-h")
        {
            std::cout
                << "Usage: " << argv[0] << " [options]\n\n"
                << "Option A – provide pre-converted files:\n"
                << "  --config <path>       config.json (from t5_generate.py)\n"
                << "  --weights <path>      weights.safetensors\n"
                << "  --encoder-ids <ids>   comma-separated encoder token IDs\n"
                << "  --decoder-ids <ids>   comma-separated decoder start IDs\n\n"
                << "Option B – let the binary call Python automatically:\n"
                << "  --model <name|path>   HF model name or local path\n"
                << "  --encoder-prompt <text>  source text\n"
                << "  --output-dir <path>   scratch directory "
                   "(default: /tmp/nntile_t5)\n\n"
                << "Common:\n"
                << "  --max-tokens <N>      tokens to generate (default: 32)\n";
            std::exit(0);
        }
    }
    return a;
}

static std::vector<std::int64_t> parse_ids(const std::string& s)
{
    std::string data = s;
    {
        std::ifstream f(s);
        if(f.good())
        {
            std::string contents((std::istreambuf_iterator<char>(f)),
                                  std::istreambuf_iterator<char>());
            if(!contents.empty())
                data = contents;
        }
    }

    std::vector<std::int64_t> ids;
    std::istringstream ss(data);
    std::string tok;
    while(std::getline(ss, tok, ','))
    {
        while(!tok.empty() && (tok.back() == '\n' || tok.back() == '\r'))
            tok.pop_back();
        if(!tok.empty())
            ids.push_back(std::stoll(tok));
    }
    return ids;
}

// ── Config loader ────────────────────────────────────────────────────────

static T5Config load_config(const std::string& path)
{
    std::ifstream f(path);
    if(!f.good())
    {
        throw std::runtime_error("Cannot open config: " + path);
    }
    json j = json::parse(f);

    T5Config cfg;
    cfg.vocab_size           = j.value("vocab_size", 32100);
    cfg.d_model              = j.value("d_model", 512);
    cfg.d_kv                 = j.value("d_kv", 64);
    cfg.d_ff                 = j.value("d_ff", 1024);
    cfg.num_layers           = j.value("num_layers", 6);
    cfg.num_decoder_layers   = j.value("num_decoder_layers", 6);
    cfg.num_heads            = j.value("num_heads", 8);
    cfg.layer_norm_epsilon   = j.value("layer_norm_epsilon", 1e-5f);
    cfg.eos_token_id         = j.value("eos_token_id", 1);
    cfg.pad_token_id         = j.value("pad_token_id", 0);
    cfg.decoder_start_token_id = j.value("decoder_start_token_id", 0);
    cfg.validate();
    return cfg;
}

// ── Weight cache ─────────────────────────────────────────────────────────

using WeightCache = std::map<std::string, std::vector<std::uint8_t>>;

static WeightCache load_weights_to_memory(const std::string& path)
{
    io::SafeTensorsReader reader(path);
    WeightCache cache;
    for(const auto& name : reader.tensor_names())
    {
        cache[name] = reader.read_tensor(name);
    }
    return cache;
}

static void apply_weight_cache(
    graph::module::Module& model,
    const WeightCache& cache)
{
    for(const auto& [name, tensor] : model.named_parameters_recursive())
    {
        auto it = cache.find(name);
        if(it != cache.end())
        {
            auto copy = it->second;
            tensor->data()->set_bind_hint(std::move(copy));
            tensor->mark_input(true);
        }
    }
}

// ── Python invocation ────────────────────────────────────────────────────

static bool run_python_conversion(const Args& args)
{
    std::vector<std::string> argv_strs = {
        "python3",
        "examples/t5_generate.py",
        "--model",
        args.model,
        "--output-dir",
        args.output_dir,
    };
    if(!args.encoder_prompt.empty())
    {
        argv_strs.push_back("--encoder-prompt");
        argv_strs.push_back(args.encoder_prompt);
    }

    std::vector<char*> argv;
    argv.reserve(argv_strs.size() + 1);
    for(auto& s : argv_strs)
        argv.push_back(s.data());
    argv.push_back(nullptr);

    std::cout << "Running: python3 examples/t5_generate.py --model "
              << args.model << " --output-dir " << args.output_dir;
    if(!args.encoder_prompt.empty())
        std::cout << " --encoder-prompt " << args.encoder_prompt;
    std::cout << "\n";

    pid_t pid = fork();
    if(pid < 0)
    {
        std::cerr << "fork() failed\n";
        return false;
    }
    if(pid == 0)
    {
        execvp("python3", argv.data());
        std::cerr << "execvp(python3) failed\n";
        _exit(127);
    }
    int status;
    if(waitpid(pid, &status, 0) != pid)
    {
        std::cerr << "waitpid() failed\n";
        return false;
    }
    return WIFEXITED(status) && WEXITSTATUS(status) == 0;
}

static std::string read_file_trimmed(const std::string& path)
{
    std::ifstream f(path);
    if(!f.good())
        return {};
    std::string s((std::istreambuf_iterator<char>(f)),
                   std::istreambuf_iterator<char>());
    while(!s.empty() && (s.back() == '\n' || s.back() == '\r'))
        s.pop_back();
    return s;
}

// ── Greedy argmax over last-position logits ──────────────────────────────

static std::int64_t argmax_last_position(
    const std::vector<float>& logits,
    Index vocab_size,
    Index seq_len)
{
    auto offset = static_cast<std::size_t>((seq_len - 1) * vocab_size);
    auto begin = logits.begin()
        + static_cast<std::ptrdiff_t>(offset);
    auto end = begin + vocab_size;
    return static_cast<std::int64_t>(std::distance(
        begin, std::max_element(begin, end)));
}

// ── Main ─────────────────────────────────────────────────────────────────

int main(int argc, char** argv)
{
    Args args = parse_args(argc, argv);

    if(!args.model.empty())
    {
        if(!run_python_conversion(args))
        {
            std::cerr << "Python conversion failed.\n";
            return 1;
        }
        args.config_path   = args.output_dir + "/config.json";
        args.weights_path  = args.output_dir + "/weights.safetensors";

        if(args.encoder_ids_str.empty())
        {
            args.encoder_ids_str = read_file_trimmed(
                args.output_dir + "/encoder_ids.txt");
        }
        if(args.decoder_ids_str.empty())
        {
            args.decoder_ids_str = read_file_trimmed(
                args.output_dir + "/decoder_ids.txt");
        }
    }

    if(args.config_path.empty() || args.weights_path.empty()
       || args.encoder_ids_str.empty() || args.decoder_ids_str.empty())
    {
        std::cerr
            << "Error: provide either (--model + --encoder-prompt) or "
               "(--config + --weights + --encoder-ids + --decoder-ids).\n"
               "Run with --help for usage.\n";
        return 1;
    }

    T5Config config = load_config(args.config_path);
    std::cout << "Config: d_model=" << config.d_model
              << "  encoder_layers=" << config.num_layers
              << "  decoder_layers=" << config.num_decoder_layers
              << "  heads=" << config.num_heads
              << "  vocab=" << config.vocab_size << "\n";

    std::vector<std::int64_t> encoder_tokens = parse_ids(args.encoder_ids_str);
    std::vector<std::int64_t> decoder_tokens = parse_ids(args.decoder_ids_str);

    if(encoder_tokens.empty())
    {
        std::cerr << "Error: encoder token list is empty.\n";
        return 1;
    }
    if(decoder_tokens.empty())
    {
        std::cerr << "Error: decoder token list is empty.\n";
        return 1;
    }
    std::cout << "Encoder: " << encoder_tokens.size() << " tokens\n";
    std::cout << "Decoder start: " << decoder_tokens.size() << " tokens\n";

    std::cout << "Loading weights from " << args.weights_path << " ...\n";
    WeightCache weights = load_weights_to_memory(args.weights_path);
    std::cout << "Cached " << weights.size() << " parameter tensors\n";

    Context context(1, 0, 0, "/tmp/nntile_ooc", 16777216, 0,
                    "localhost", 5001, 0);

    std::cout << "\n--- Generating (greedy) ---\n";
    for(int step = 0; step < args.max_tokens; ++step)
    {
        Index enc_seq = static_cast<Index>(encoder_tokens.size());
        Index dec_seq = static_cast<Index>(decoder_tokens.size());

        NNGraph graph("t5_step");

        auto* encoder_ids = graph.tensor(
            {enc_seq, 1}, "encoder_input_ids", DataType::INT64, false);
        encoder_ids->mark_input(true);

        auto* decoder_ids = graph.tensor(
            {dec_seq, 1}, "decoder_input_ids", DataType::INT64, false);
        decoder_ids->mark_input(true);

        T5ForConditionalGeneration model(&graph, "model", config);
        auto* output = model.forward(encoder_ids, decoder_ids);
        output->mark_output(true);

        apply_weight_cache(model, weights);

        TensorGraph::Runtime runtime(graph.tensor_graph());
        runtime.compile();
        runtime.bind_data("encoder_input_ids", encoder_tokens);
        runtime.bind_data("decoder_input_ids", decoder_tokens);
        runtime.execute();
        runtime.wait();

        auto logits = runtime.get_output<float>(output->name());

        std::int64_t next_id = argmax_last_position(
            logits, config.vocab_size, dec_seq);
        decoder_tokens.push_back(next_id);

        std::cout << "step " << (step + 1) << "  token=" << next_id << "\n";

        if(next_id == config.eos_token_id)
        {
            std::cout << "EOS reached.\n";
            break;
        }
    }

    std::cout << "\nGenerated decoder token IDs:";
    for(auto id : decoder_tokens)
    {
        std::cout << " " << id;
    }
    std::cout << "\n";

    std::cout << "\nTo decode, run:\n"
              << "  python3 -c \"from transformers import AutoTokenizer; "
              << "t = AutoTokenizer.from_pretrained('<model-dir>'); "
              << "print(t.decode([";
    for(std::size_t i = 0; i < decoder_tokens.size(); ++i)
    {
        if(i > 0) std::cout << ",";
        std::cout << decoder_tokens[i];
    }
    std::cout << "]))\"\n";

    return 0;
}

File: examples/t5_generate.py
#!/usr/bin/env python3
# @copyright (c) 2022-present Skolkovo Institute of Science and Technology
#                              (Skoltech), Russia. All rights reserved.
#                2023-present Artificial Intelligence Research Institute
#                              (AIRI), Russia. All rights reserved.
#
# NNTile is software framework for fast training of big neural networks on
# distributed-memory heterogeneous systems based on StarPU runtime system.
#
# @file examples/t5_generate.py
# Convert a HuggingFace T5 checkpoint to NNTile format for C++ inference.
#
# Usage (requires T5 v1.1+ gated models, e.g. flan-t5-small):
#   python examples/t5_generate.py \
#       --model google/flan-t5-small \
#       --output-dir /tmp/nntile_t5 \
#       --encoder-prompt "translate English to German: The house is wonderful."
#
# Then run the C++ binary:
#   ./t5_generate \
#       --config /tmp/nntile_t5/config.json \
#       --weights /tmp/nntile_t5/weights.safetensors \
#       --encoder-ids "$(cat /tmp/nntile_t5/encoder_ids.txt)" \
#       --decoder-ids "$(cat /tmp/nntile_t5/decoder_ids.txt)" \
#       --max-tokens 32
#
# @version 1.1.0

"""Convert HuggingFace T5 checkpoint to NNTile format.

Produces:
  - weights.safetensors  NNTile-layout weight file
  - config.json  Model configuration readable by the C++ example
  - encoder_ids.txt  Comma-separated token IDs for encoder input
  - decoder_ids.txt  Comma-separated token IDs for decoder start
"""

from __future__ import annotations

import argparse
import json
import struct
import sys
from pathlib import Path
from typing import Callable, Sequence

import numpy as np
import torch
from huggingface_hub import snapshot_download
from safetensors import safe_open
from transformers import AutoConfig, AutoTokenizer

# ── Layout helpers ────────────────────────────────────────────────────────


def fortran_order(arr: np.ndarray) -> np.ndarray:
    """Return C-contiguous array whose flat bytes equal NNTile column-major."""
    a = np.asarray(arr, dtype=np.float32)
    return a.ravel("F").reshape(a.shape)


# ── Streaming safetensors writer ──────────────────────────────────────────


def _write_safetensors_streaming(
    path: str | Path,
    specs: Sequence[tuple[str, tuple[int, ...]]],
    get_data: Callable[[str], np.ndarray],
) -> None:
    """Write a safetensors file one tensor at a time."""
    header: dict[str, dict] = {}
    offset = 0
    entry_order: list[tuple[str, int]] = []
    for name, shape in specs:
        nbytes = int(np.prod(shape)) * 4
        header[name] = {
            "dtype": "F32",
            "shape": list(shape),
            "data_offsets": [offset, offset + nbytes],
        }
        entry_order.append((name, nbytes))
        offset += nbytes

    header_bytes = json.dumps(header, separators=(",", ":")).encode("utf-8")
    pad = (8 - len(header_bytes) % 8) % 8
    header_bytes += b" " * pad

    with open(path, "wb") as f:
        f.write(struct.pack("<Q", len(header_bytes)))
        f.write(header_bytes)
        for name, nbytes in entry_order:
            arr = get_data(name)
            raw = arr.tobytes()
            assert len(raw) == nbytes, (
                f"{name}: expected {nbytes} bytes, got {len(raw)}")
            f.write(raw)
            del arr, raw


# ── Output tensor specs (NNTile parameter names) ──────────────────────────


def _output_specs(config) -> list[tuple[str, tuple[int, ...]]]:
    """Return (nntile_name, shape) for every output tensor."""
    d_model = config.d_model
    d_ff = config.d_ff
    V = config.vocab_size
    n_heads = config.num_heads
    d_kv = getattr(config, "d_kv", d_model // n_heads)  # head_size

    specs: list[tuple[str, tuple[int, ...]]] = []

    # Shared embedding
    specs.append(("model.model.embed_tokens.vocab", (d_model, V)))

    # Encoder
    for i in range(config.num_layers):
        p = f"model.model.encoder_layers_{i}"
        specs.append((f"{p}.layer_norm_0.gamma", (d_model,)))
        specs.append((f"{p}.self_attn.q_weight", (n_heads, d_kv, d_model)))
        specs.append((f"{p}.self_attn.k_weight", (n_heads, d_kv, d_model)))
        specs.append((f"{p}.self_attn.v_weight", (n_heads, d_kv, d_model)))
        specs.append((f"{p}.self_attn.o_weight", (d_model, n_heads, d_kv)))
        specs.append((f"{p}.ff.layer_norm.gamma", (d_model,)))
        specs.append((f"{p}.ff.dense.gate_proj.weight", (d_model, d_ff)))
        specs.append((f"{p}.ff.dense.up_proj.weight", (d_model, d_ff)))
        specs.append((f"{p}.ff.dense.down_proj.weight", (d_ff, d_model)))

    specs.append(("model.model.encoder_final_norm.gamma", (d_model,)))

    # Decoder
    for i in range(config.num_decoder_layers):
        p = f"model.model.decoder_layers_{i}"
        specs.append((f"{p}.layer_norm_0.gamma", (d_model,)))
        specs.append((f"{p}.self_attn.q_weight", (n_heads, d_kv, d_model)))
        specs.append((f"{p}.self_attn.k_weight", (n_heads, d_kv, d_model)))
        specs.append((f"{p}.self_attn.v_weight", (n_heads, d_kv, d_model)))
        specs.append((f"{p}.self_attn.o_weight", (d_model, n_heads, d_kv)))
        specs.append((f"{p}.layer_norm_1.gamma", (d_model,)))
        specs.append((f"{p}.cross_attn.q_weight", (n_heads, d_kv, d_model)))
        specs.append((f"{p}.cross_attn.k_weight", (n_heads, d_kv, d_model)))
        specs.append((f"{p}.cross_attn.v_weight", (n_heads, d_kv, d_model)))
        specs.append((f"{p}.cross_attn.o_weight", (d_model, n_heads, d_kv)))
        specs.append((f"{p}.ff.layer_norm.gamma", (d_model,)))
        specs.append((f"{p}.ff.dense.gate_proj.weight", (d_model, d_ff)))
        specs.append((f"{p}.ff.dense.up_proj.weight", (d_model, d_ff)))
        specs.append((f"{p}.ff.dense.down_proj.weight", (d_ff, d_model)))

    specs.append(("model.model.decoder_final_norm.gamma", (d_model,)))

    # LM head
    specs.append(("model.lm_head.weight", (d_model, V)))

    return specs


# ── Per-tensor conversion ──────────────────────────────────────────────────


def _make_converter(
    config,
    hf_get: Callable[[str], np.ndarray],
    has_tensor: Callable[[str], bool],
) -> Callable[[str], np.ndarray]:
    """Return a function that converts a single NNTile tensor on demand."""
    d_model = config.d_model
    n_heads = config.num_heads
    d_kv = getattr(config, "d_kv", d_model // n_heads)

    def _get_ff_wi_2(hp: str, layer_idx: int) -> np.ndarray:
        key = f"{hp}.layer.{layer_idx}.DenseReluDense.wi_2.weight"
        if not has_tensor(key):
            raise KeyError(
                "T5 v1.0 (non-gated) models are not supported. "
                f"Missing {key}. Use T5 v1.1+ (e.g. google/flan-t5-small)."
            )
        return hf_get(key)

    def convert(name: str) -> np.ndarray:
        if name == "model.model.embed_tokens.vocab":
            return fortran_order(hf_get("shared.weight").T)

        if name == "model.model.encoder_final_norm.gamma":
            return fortran_order(hf_get("encoder.final_layer_norm.weight"))

        if name == "model.model.decoder_final_norm.gamma":
            return fortran_order(hf_get("decoder.final_layer_norm.weight"))

        if name == "model.lm_head.weight":
            return fortran_order(hf_get("lm_head.weight").T)

        # Encoder layers
        if "encoder_layers_" in name and "decoder" not in name:
            parts = name.split(".")
            layer_idx = int(parts[2].split("_", 2)[2])
            rest = ".".join(parts[3:])
            hp = f"encoder.block.{layer_idx}"

            if rest == "layer_norm_0.gamma":
                return fortran_order(
                    hf_get(f"{hp}.layer.0.layer_norm.weight"))
            if rest == "self_attn.q_weight":
                q = hf_get(f"{hp}.layer.0.SelfAttention.q.weight")
                return fortran_order(
                    q.reshape(n_heads, d_kv, d_model))
            if rest == "self_attn.k_weight":
                k = hf_get(f"{hp}.layer.0.SelfAttention.k.weight")
                return fortran_order(k.reshape(n_heads, d_kv, d_model))
            if rest == "self_attn.v_weight":
                v = hf_get(f"{hp}.layer.0.SelfAttention.v.weight")
                return fortran_order(v.reshape(n_heads, d_kv, d_model))
            if rest == "self_attn.o_weight":
                o = hf_get(f"{hp}.layer.0.SelfAttention.o.weight")
                return fortran_order(o.reshape(d_model, n_heads, d_kv))

            if rest == "ff.layer_norm.gamma":
                return fortran_order(
                    hf_get(f"{hp}.layer.1.layer_norm.weight"))
            if rest == "ff.dense.gate_proj.weight":
                return fortran_order(
                    hf_get(f"{hp}.layer.1.DenseReluDense.wi.weight").T)
            if rest == "ff.dense.up_proj.weight":
                return fortran_order(
                    _get_ff_wi_2(hp, 1).T)
            if rest == "ff.dense.down_proj.weight":
                return fortran_order(
                    hf_get(f"{hp}.layer.1.DenseReluDense.wo.weight").T)

        # Decoder layers
        if "decoder_layers_" in name:
            parts = name.split(".")
            layer_idx = int(parts[2].split("_", 2)[2])
            rest = ".".join(parts[3:])
            hp = f"decoder.block.{layer_idx}"

            if rest == "layer_norm_0.gamma":
                return fortran_order(
                    hf_get(f"{hp}.layer.0.layer_norm.weight"))
            if rest == "self_attn.q_weight":
                q = hf_get(f"{hp}.layer.0.SelfAttention.q.weight")
                return fortran_order(q.reshape(n_heads, d_kv, d_model))
            if rest == "self_attn.k_weight":
                k = hf_get(f"{hp}.layer.0.SelfAttention.k.weight")
                return fortran_order(k.reshape(n_heads, d_kv, d_model))
            if rest == "self_attn.v_weight":
                v = hf_get(f"{hp}.layer.0.SelfAttention.v.weight")
                return fortran_order(v.reshape(n_heads, d_kv, d_model))
            if rest == "self_attn.o_weight":
                o = hf_get(f"{hp}.layer.0.SelfAttention.o.weight")
                return fortran_order(o.reshape(d_model, n_heads, d_kv))

            if rest == "layer_norm_1.gamma":
                return fortran_order(
                    hf_get(f"{hp}.layer.1.layer_norm.weight"))
            if rest == "cross_attn.q_weight":
                q = hf_get(f"{hp}.layer.1.EncDecAttention.q.weight")
                return fortran_order(q.reshape(n_heads, d_kv, d_model))
            if rest == "cross_attn.k_weight":
                k = hf_get(f"{hp}.layer.1.EncDecAttention.k.weight")
                return fortran_order(k.reshape(n_heads, d_kv, d_model))
            if rest == "cross_attn.v_weight":
                v = hf_get(f"{hp}.layer.1.EncDecAttention.v.weight")
                return fortran_order(v.reshape(n_heads, d_kv, d_model))
            if rest == "cross_attn.o_weight":
                o = hf_get(f"{hp}.layer.1.EncDecAttention.o.weight")
                return fortran_order(o.reshape(d_model, n_heads, d_kv))

            if rest == "ff.layer_norm.gamma":
                return fortran_order(
                    hf_get(f"{hp}.layer.2.layer_norm.weight"))
            if rest == "ff.dense.gate_proj.weight":
                return fortran_order(
                    hf_get(f"{hp}.layer.2.DenseReluDense.wi.weight").T)
            if rest == "ff.dense.up_proj.weight":
                return fortran_order(
                    _get_ff_wi_2(hp, 2).T)
            if rest == "ff.dense.down_proj.weight":
                return fortran_order(
                    hf_get(f"{hp}.layer.2.DenseReluDense.wo.weight").T)

        raise ValueError(f"Unknown NNTile tensor: {name}")

    return convert


# ── Main ──────────────────────────────────────────────────────────────────


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Convert HF T5 checkpoint → NNTile format",
    )
    parser.add_argument(
        "--model", required=True,
        help="HF model (e.g. google/flan-t5-small) or local path",
    )
    parser.add_argument(
        "--output-dir", required=True,
        help="Directory where NNTile files will be written",
    )
    parser.add_argument(
        "--encoder-prompt", default=None,
        help="Optional text for encoder input (source sequence)",
    )
    parser.add_argument(
        "--decoder-prompt", default=None,
        help="Optional text for decoder start (default: decoder_start_token)",
    )
    args = parser.parse_args()

    model_id = args.model

    if Path(model_id).is_dir():
        model_dir = Path(model_id)
    else:
        print(f"Downloading {model_id} to HF cache ...")
        model_dir = Path(snapshot_download(model_id))
        print(f"Snapshot: {model_dir}")

    out_dir = Path(args.output_dir)
    out_dir.mkdir(parents=True, exist_ok=True)

    config = AutoConfig.from_pretrained(str(model_dir))
    is_gated = getattr(config, "is_gated_act", True)
    if not is_gated:
        print(
            "ERROR: T5 v1.0 (non-gated) models are not supported. "
            "Use T5 v1.1+ gated models (e.g. google/flan-t5-small).",
            file=sys.stderr,
        )
        return 1
    print(f"Model: {config.model_type}  d_model={config.d_model}  "
          f"layers={config.num_layers}/{config.num_decoder_layers}  "
          f"vocab={config.vocab_size}")

    num_decoder_layers = getattr(
        config, "num_decoder_layers", config.num_layers)
    nntile_config = {
        "vocab_size": config.vocab_size,
        "d_model": config.d_model,
        "d_kv": config.d_kv,
        "d_ff": config.d_ff,
        "num_layers": config.num_layers,
        "num_decoder_layers": num_decoder_layers,
        "num_heads": config.num_heads,
        "layer_norm_epsilon": getattr(config, "layer_norm_epsilon", 1e-5),
        "eos_token_id": getattr(config, "eos_token_id", 1),
        "pad_token_id": getattr(config, "pad_token_id", 0),
        "decoder_start_token_id": getattr(config, "decoder_start_token_id", 0),
    }
    config_path = out_dir / "config.json"
    config_path.write_text(json.dumps(nntile_config, indent=2) + "\n")
    print(f"Wrote {config_path}")

    st_files = sorted(model_dir.glob("*.safetensors"))
    if not st_files:
        print(f"ERROR: no *.safetensors files found in {model_dir}",
              file=sys.stderr)
        return 1

    handles: list = []
    tensor_to_handle: dict[str, object] = {}
    for sf in st_files:
        h = safe_open(str(sf), framework="pt")
        handles.append(h)
        for key in h.keys():
            tensor_to_handle[key] = h

    tth = tensor_to_handle

    def hf_get(name: str) -> np.ndarray:
        return (tth[name].get_tensor(name).to(torch.float32).numpy())

    def has_tensor(name: str) -> bool:
        return name in tth

    specs = _output_specs(config)
    converter = _make_converter(config, hf_get, has_tensor)

    weights_path = out_dir / "weights.safetensors"
    print(f"Converting {len(specs)} tensors (streaming) ...")
    _write_safetensors_streaming(weights_path, specs, converter)
    print(f"Wrote {weights_path}")

    del handles, tensor_to_handle

    tokenizer = AutoTokenizer.from_pretrained(str(model_dir))

    if args.encoder_prompt is not None:
        enc_ids = tokenizer.encode(
            args.encoder_prompt,
            add_special_tokens=True,
            return_tensors="np",
        ).flatten()
        enc_str = ",".join(str(i) for i in enc_ids)
        (out_dir / "encoder_ids.txt").write_text(enc_str + "\n")
        print(f"Encoder prompt ({len(enc_ids)} tokens): {enc_ids.tolist()}")
        print(f"Wrote {out_dir / 'encoder_ids.txt'}")

    if args.decoder_prompt is not None:
        dec_ids = tokenizer.encode(
            args.decoder_prompt,
            add_special_tokens=False,
            return_tensors="np",
        ).flatten()
        dec_str = ",".join(str(i) for i in dec_ids)
        (out_dir / "decoder_ids.txt").write_text(dec_str + "\n")
        print(f"Decoder prompt ({len(dec_ids)} tokens): {dec_ids.tolist()}")
    else:
        dec_start = config.decoder_start_token_id
        if dec_start is None:
            dec_start = config.pad_token_id
        (out_dir / "decoder_ids.txt").write_text(str(dec_start) + "\n")
        print(f"Decoder start: {dec_start}")

    print("Done.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
